staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -778,22 +779,47 @@ func (r *Request) newStreamWatcher(resp *http.Response) (watch.Interface, error)
 	), nil
 }
 
-// updateURLMetrics is a convenience function for pushing metrics.
-// It also handles corner cases for incomplete/invalid request data.
-func updateURLMetrics(ctx context.Context, req *Request, resp *http.Response, err error) {
-	url := "none"
+// updateRequestResultMetric increments the RequestResult metric counter,
+// it should be called with the (response, err) tuple from the final
+// reply from the server.
+func updateRequestResultMetric(ctx context.Context, req *Request, resp *http.Response, err error) {
+	code, host := sanitize(req, resp, err)
+	metrics.RequestResult.Increment(ctx, code, req.verb, host)
+}
+
+// updateRequestRetryMetric increments the RequestRetry metric counter,
+// it should be called with the (response, err) tuple for each retry
+// except for the final attempt.
+func updateRequestRetryMetric(ctx context.Context, req *Request, resp *http.Response, err error) {
+	code, host := sanitize(req, resp, err)
+	metrics.RequestRetry.IncrementRetry(ctx, code, req.verb, host)
+}
+
+func updateRequestRetryOrResultMetric(ctx context.Context, done bool, req *Request, resp *http.Response, err error) {
+	if !done {
+		// this is not the final attempt, there is at least one more
+		// attempt, so count this attempt as a retry.
+		updateRequestRetryMetric(ctx, req, resp, err)
+	}
+	// we count the retry attempts into the result metric, in
+	// keeping with the current behavior.
+	updateRequestResultMetric(ctx, req, resp, err)
+}
+
+func sanitize(req *Request, resp *http.Response, err error) (string, string) {
+	host := "none"
```

Comment:
I would prefer defaulting to an empty string, it should be more straightforward for the users to eliminate metrics with an empty host by doing `{host!=""}` rather than `{host!="none"}`.

Reply:
this will change the behavior of the other client metrics too since `sanitize` function is used by other client metrics. I think we should change the default in a separate PR, thoughts?

Reply:
make sense to me :+1:

Reply:
I opened https://github.com/kubernetes/kubernetes/issues/115461